### PR TITLE
fix: bug in passing Versioned field set for getHealReplicationInfo()

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: bug in passing Versioned field set for getHealReplicationInfo()

## Motivation and Context
Bonus:

- rejects prefix deletes on object-locked buckets earlier

## How to test this PR?
CI/CD should cover all the tests

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -2293,7 +2293,12 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
 		return
 	}
 
-	if s3Error := checkRequestAuthType(ctx, r, policy.DeleteObjectAction, bucket, object); s3Error != ErrNone {
+	replica := r.Header.Get(xhttp.AmzBucketReplicationStatus) == replication.Replica.String()
+	action := policy.DeleteObjectAction
+	if replica {
+		action = policy.ReplicateDeleteAction
+	}
+	if s3Error := checkRequestAuthType(ctx, r, policy.Action(action), bucket, object); s3Error != ErrNone {
```
Comment (klauspost) [inline cmd/object-handlers.go]: AFAICT nothing prevents someone from just setting this property and change the check?

It seems to me we really needs some documentation to make sure that replica users are never mixed with regular users.

Comment (harshavardhana) [reply cmd/object-handlers.go]: This change is just relaxing the policy requirement, currently it requires both DeleteObject() and ReplicateDelete()

For replicated request we only need ReplicateDelete() that's all

File: cmd/object-handlers.go [hunk 2]
```diff
@@ -2338,17 +2345,10 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
...
 	vID := opts.VersionID
-	if r.Header.Get(xhttp.AmzBucketReplicationStatus) == replication.Replica.String() {
-		// check if replica has permission to be deleted.
-		if apiErrCode := checkRequestAuthType(ctx, r, policy.ReplicateDeleteAction, bucket, object); apiErrCode != ErrNone {
-			writeErrorResponse(ctx, w, errorCodes.ToAPIErr(apiErrCode), r.URL)
-			return
-		}
+	if dsc.ReplicateAny() {
+		opts.SetDeleteReplicationState(dsc, opts.VersionID)
 		opts.SetReplicaStatus(replication.Replica)
```
Comment (poornas) [inline cmd/object-handlers.go:2357]: replica status should not be set on the replicating side. Should be set only if replica bool is set to true

**Review by poornas — CHANGES_REQUESTED** [review]

File: cmd/object-handlers.go [hunk 3]
```diff
@@ -2338,42 +2345,31 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
...
-		opts.SetReplicaStatus(replication.Replica)
-		if opts.VersionPurgeStatus().Empty() {
-			// opts.VersionID holds delete marker version ID to replicate and not yet present on disk
-			vID = ""
+	if dsc.ReplicateAny() {
+		opts.SetDeleteReplicationState(dsc, opts.VersionID)
+		if replica {
+			opts.SetReplicaStatus(replication.Replica)
+			if opts.VersionPurgeStatus().Empty() {
+				// opts.VersionID holds delete marker version ID to replicate and not yet present on disk
```
Comment (poornas) [inline cmd/object-handlers.go]: this should be outside the dsc.ReplicateAny() - on replica side, replication decision will be false

File: cmd/object-handlers.go (hunk 1 above)
Comment (klauspost) [reply cmd/object-handlers.go]: Yeah, but my point is that you can change the permission required from `DeleteObjectAction` to `ReplicateDeleteAction` by setting a header in the request.

If users are somehow given `ReplicateDeleteAction`, a hostile user could delete objects without `DeleteObjectAction` permission.

Comment (harshavardhana) [reply cmd/object-handlers.go]: if that confusion can arise, I can simply remove this.

**Comment by harshavardhana** [comment]
I removed the permissions back to what it was to avoid any confusion. We may have to add replica-request as part of the signed headers to avoid spoofing.

**Comment by harshavardhana** [comment]
PTAL @poornas @vadmeste @klauspost

**Review by klauspost — APPROVED** [review]
LGTM

**Comment by harshavardhana** [comment]
PTAL @poornas

**Review by poornas — APPROVED** [review]

**Pull request merged**